### ISSUE TITLE
RandR Deserialisation Error

### ISSUE THREAD

**Issue by thedward** [aslatter/xhb#13]
I am getting an error when I try to access the returned `Graphics.XHB.Gen.RandR.GetScreenInfoReply` value.

When I try and run my program, I get the following output:

```
HAS XRANDR!
jrandr: Data.Binary.Get.runGet at position 72: not enough bytes
```

I am using version `0.6.2015.8.1`.

Here is my program:

``` haskell
module Main where

import Foreign.C.String
import qualified Data.ByteString.Lazy as BS
import Control.Concurrent
import Control.Monad

import qualified Graphics.XHB as X
import qualified Graphics.XHB.Gen.RandR as R

import System.IO

main = do
  connectionM <- X.connect

  case connectionM of
    Nothing -> putStrLn "failed to get connection"
    Just c -> randr c

randr :: X.Connection -> IO ()
randr c = do

  listReceipt <- X.listExtensions c

  replyOrError <- X.getReply listReceipt
  let extensions = case replyOrError of
                        Left e -> error $ "error in extensions request" ++ show e
                        Right listRep ->  listExtensionNames listRep

  screenInfoReceipt <-
    if "RANDR" `elem` extensions
       then putStrLn "HAS XRANDR!" >> R.getScreenInfo c (X.getRoot c)
       else error "NO XRANDR :-("

  replyOrError <- X.getReply screenInfoReceipt
  let info = case replyOrError of
                  Left e -> error (show e)
                  Right info -> info

  print info

listExtensionNames :: X.ListExtensionsReply -> [String]
listExtensionNames = map strToString . X.names_ListExtensionsReply

strToString :: X.STR -> String
strToString = map castCCharToChar . X.name_STR
```

**Comment by adrianparvino** [comment]
A whole year and a few months late. You need to advertise your version first, since XHB doesn't take account for versioning.

**Comment by thedward** [comment]
@adrianparvino , can you elaborate on what you mean by "advertise your version"?

**Comment by adrianparvino** [comment]
`RandR.queryVersion connection 1 2`

That will tell the X server that you accept randr replies with version 1.2. Turns out that 1.0 is not forward compatible with 1.2.

**Comment by thedward** [comment]
@adrianparvino,

Thank you! That did the trick.

**Comment by thedward** [comment]
There is now a different — but similar — error:

     Data.Binary.Get.runGet at position 124: not enough bytes 

This particular error only occurs when two displays are connected and one of them is turned off — it works fine when both of the displays are on.

@adrianparvino,

How did you know/find the problem for the previous error?  I'm not opposed to digging through source, but I would be tremendously grateful if you could point me in the right direction.

**Comment by adrianparvino** [comment]
@thedward No problem. The more people that gets to checking this, the better. I checked the [randr specs](https://cgit.freedesktop.org/xorg/proto/randrproto/tree/randrproto.txt) and the [randr implementation on the x server](https://github.com/mirror/xserver/tree/master/randr). I have a feeling that the X server is non-compliant with the specs, but that's not my call.

I also unpacked the xhb source code (since I had to make some revisions on the .cabal file too). Files found [here](https://github.com/aslatter/xhb/releases).

**Comment by thedward** [comment]
I think I might have found at least part of the problem.

The definition of [`RRGetScreenInfo` in the protocol documentation](https://cgit.freedesktop.org/xorg/proto/randrproto/tree/randrproto.txt#n2434) specifies a field that
contains the total size of the refresh rate data in 2-byte units. However, the [XCB XML specification](https://cgit.freedesktop.org/xcb/proto/tree/src/randr.xml#n138) of the protocol — and thus XHB — specifies that the *count* of refresh rate records can be calculated by subtracting the number of screen size records from the refresh rate data size; based on my current understanding this will only be true if each refresh rate record only contains a single rate (otherwise xhb will try and parse more records than are there).

**Comment by adrianparvino** [comment]
@thedward That was also the part that I considered non-compliant. I should try to make a patch.

**Comment by thedward** [comment]
@adrianparvino,

Since [randrproto.txt](https://cgit.freedesktop.org/xorg/proto/randrproto/tree/randrproto.txt#n602) says — emphasis mine:

> 'refresh' is the list of refresh rates for each size. __*Each element of 'sizes' has a corresponding element in 'refresh'.*__ An empty list indicates no known rates, or a device for which refresh is not relevant.

I think this could be fixed on the XCB-XML side with something as simple as:
		
```diff		
diff --git a/src/randr.xml b/src/randr.xml
index 722f715..e7cea84 100644
--- a/src/randr.xml
+++ b/src/randr.xml
@@ -152,10 +152,7 @@ authorization from the authors.
 			    <fieldref>nSizes</fieldref>
 			</list>
 			<list type="RefreshRates" name="rates">
-				<op op="-">
-					<fieldref>nInfo</fieldref>
-					<fieldref>nSizes</fieldref>
-				</op>
+			    <fieldref>nSizes</fieldref>
 			</list>
 		</reply>
 	</request>
```

What do you think?

**Comment by adrianparvino** [comment]
That actually seems correct. Do you have a local version of XHB? I think it'd be easier to patch XHB than XCB-XML. If it's actually correct, then we can send a patch to XCB.

**Comment by thedward** [comment]
Yep, making that fix in the generated Haskell source fixes my problem.

I made the following change:

```diff
--- a/patched/Graphics/XHB/Gen/RandR/Types.hs    2016-07-27 10:09:47.711984702 -0500
+++ b/patched/Graphics/XHB/Gen/RandR/Types.hs    2017-06-19 15:54:59.819842149 -0500
@@ -315,7 +315,7 @@
                skip 2
                sizes <- deserializeList (fromIntegral nSizes)
                rates <- deserializeList
-                          (fromIntegral (fromIntegral (nInfo - nSizes)))
+                          (fromIntegral nSizes)
                let _ = isCard32 length
                return
                  (MkGetScreenInfoReply rotations root timestamp config_timestamp
```
